### PR TITLE
[CBRD-23159] Updated slave node GCM

### PR DESCRIPTION
**Pull request by alexstanro**
http://jira.cubrid.org/browse/CBRD-23159

Added group complete class for slave node and related.

For compile reason, some code in CBRD-23158 (master node) repeats in current patch, so it may be skipped from CR.

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/connection/server_support.c (258 incorrect indentation(s) found)
```
  2485   int css_notify_ha_log_applier_state (THREAD_ENTRY * thread_p, HA_LOG_APPLIER_STATE state)
  2486   {
  2487     HA_LOG_APPLIER_STATE_TABLE *table;
  2488     HA_SERVER_STATE server_state;
  2489     int i, client_id;
  2490 
  2491     assert (state >= HA_LOG_APPLIER_STATE_UNREGISTERED && state <= HA_LOG_APPLIER_STATE_ERROR);
  2492 
  2493     csect_enter (thread_p, CSECT_HA_SERVER_STATE, INF_WAIT);
  2494 
  2495     client_id = css_get_client_id (thread_p);
  2496     er_log_debug (ARG_FILE_LINE, "css_notify_ha_log_applier_state: client %d state %s\n", client_id,
  2497 		  css_ha_applier_state_string (state));
  2498     for (i = 0, table = ha_Log_applier_state; i < ha_Log_applier_state_num; i++, table++)
  2499       {
  2500 	if (table->client_id == client_id)
  2501 	  {
  2502 	    if (table->state == state)
  2503 	      {
  2504 		csect_exit (thread_p, CSECT_HA_SERVER_STATE);
```
... More than 20 lines are omitted.
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/replication/log_consumer.hpp [hunk 1]
```diff
@@ -163,6 +163,17 @@ namespace cubreplication
       subtran_applier &get_subtran_applier ();
   };
 
+  //
+  // dispatch_consumer is the common interface used by dispatch to control group creation.
+  //
+  class dispatch_consumer
```
Comment (ribram) [inline src/replication/log_consumer.hpp]: 1. Does this class have any meaning outside slave group control manager? The virtual interface and multiple inheritance of slave GCM makes it undesirable unless there are strong arguments to keep the class separate from slave GCM.

2. In case we do need the separate class, its name conflicts with the dispatcher of tasks. Is there a better name?

File: src/connection/server_support.c [hunk 2]
```diff
@@ -2257,7 +2258,7 @@ void
 css_finish_transit (THREAD_ENTRY * thread_p, bool force, HA_SERVER_STATE req_state)
 {
   assert (req_state == HA_SERVER_STATE_ACTIVE || req_state == HA_SERVER_STATE_STANDBY);
-
+// *INDENT-OFF*
```
Comment (ribram) [inline src/connection/server_support.c]: looks like a merge issue; please fix.

File: src/transaction/log_manager.h [hunk 3]
```diff
@@ -247,4 +247,13 @@ void LOG_CS_PROMOTE (THREAD_ENTRY * thread_p);
 bool LOG_CS_OWN (THREAD_ENTRY * thread_p);
 bool LOG_CS_OWN_WRITE_MODE (THREAD_ENTRY * thread_p);
 
+// *INDENT-OFF*
+/* Interface used by log flush to notify threads. */
+class log_flush_lsa
```
Comment (ribram) [inline src/transaction/log_manager.h]: it seems log_flush_lsa class is never used by slave GCM. maybe it belongs to another patch?

File: src/transaction/transaction_group_complete_manager.cpp [hunk 4]
```diff
@@ -438,7 +466,7 @@ namespace cubtx
       }
     else
       {
-	/* Current closed group - check whether MVCC was completed. */
+	/* Current closed group - check whether MVCC was completed.*/
```
Comment (ribram) [inline src/transaction/transaction_group_complete_manager.cpp]: revert removed spaces before `*/` (same for several places)

File: src/transaction/transaction_group_complete_manager.hpp [hunk 5]
```diff
@@ -72,6 +72,9 @@ namespace cubtx
       void complete_logging (id_type group_id) override final;
 
     protected:
+      id_type set_current_group_minimum_transactions (unsigned int count_minimum_transactions,
```
Comment (ribram) [inline src/transaction/transaction_group_complete_manager.hpp]: Let's summarize discussion postponed from previous patch:

Me and razvan argued it is better to move set_current_group_minimum_transactions/get_current_group_minimum_transactions from common virtual interface to salve GCM.

It is impossible to expect a minimum number of transactions in master or single GCM; it could block a smaller number of transactions for an unpredictable period of time.

To clearly understand the purpose of these functions, it is better to move it them to their usage context. Also, it seems like slave GCM expects an exact number of transactions rather than a minimum number.

File: src/transaction/transaction_master_group_complete_manager.cpp [hunk 6]
```diff
@@ -41,10 +41,11 @@ namespace cubtx
   {
     assert (gl_master_group != NULL);
     return gl_master_group;
+
   }
 
   //
-  // init initializes master group complete
+  // init initialize master group commit
```
Comment (ribram) [inline src/transaction/transaction_master_group_complete_manager.cpp]: left comments look better.

File: src/transaction/transaction_slave_group_complete_manager.hpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      std::atomic<cubstream::stream_position> m_latest_group_stream_position;
+
+      /* m_has_latest_group_close_info - true, if stream position and count expected transactions were set. */
+      std::atomic<bool> m_has_latest_group_close_info;
+  };
+
+  //
+  // slave_group_complete_task is class for slave group complete daemon
+  //
+  class slave_group_complete_task : public cubthread::entry_task
```
Comment (ribram) [inline src/transaction/transaction_slave_group_complete_manager.hpp]: Move class to cpp files; it can also be completely replaced by cubthread::entry_callable_task.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/connection/server_support.c (258 incorrect indentation(s) found)
```
  2485   int css_notify_ha_log_applier_state (THREAD_ENTRY * thread_p, HA_LOG_APPLIER_STATE state)
  2486   {
  2487     HA_LOG_APPLIER_STATE_TABLE *table;
  2488     HA_SERVER_STATE server_state;
  2489     int i, client_id;
  2490 
  2491     assert (state >= HA_LOG_APPLIER_STATE_UNREGISTERED && state <= HA_LOG_APPLIER_STATE_ERROR);
  2492 
  2493     csect_enter (thread_p, CSECT_HA_SERVER_STATE, INF_WAIT);
  2494 
  2495     client_id = css_get_client_id (thread_p);
  2496     er_log_debug (ARG_FILE_LINE, "css_notify_ha_log_applier_state: client %d state %s\n", client_id,
  2497 		  css_ha_applier_state_string (state));
  2498     for (i = 0, table = ha_Log_applier_state; i < ha_Log_applier_state_num; i++, table++)
  2499       {
  2500 	if (table->client_id == client_id)
  2501 	  {
  2502 	    if (table->state == state)
  2503 	      {
  2504 		csect_exit (thread_p, CSECT_HA_SERVER_STATE);
```
... More than 20 lines are omitted.
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/replication/log_consumer.hpp (hunk 1 above)
Comment (alexstanro) [reply src/replication/log_consumer.hpp]: dispatch_consumer is a interface implemented by slave manager, used by consumer to notify slave manager. Please, specify a better name. Also whether you think is better to move it in other file.

File: src/transaction/log_manager.h (hunk 3 above)
Comment (alexstanro) [reply src/transaction/log_manager.h]: It is used in next patch - 23160.

File: src/replication/log_consumer.hpp (hunk 1 above)
Comment (razvanradulescu) [reply src/replication/log_consumer.hpp]: The instantiation of dispatch_consumer  is :
`m_p_dispatch_consumer = cubtx::slave_group_complete_manager::get_instance ();`

If it would be something like :
`m_p_dispatch_consumer = cubtx::complete_manager::get_instance ();`

The interface would be required.
With current code, the methods from this interface can be used only with slave GCM.

Comment (alexstanro) [reply src/replication/log_consumer.hpp]: Again, with the interface, the consumer can calls only methods that he needs. The consumer can't call other public methods from slave manager, to not corrupt the system (the consumer can't call do_complete for instance). We can move the interface in other file, rename it and I will do these changes.

Comment (ribram) [reply src/replication/log_consumer.hpp]: I understand. Still virtual interface and multiple inheritance remain; these can be avoided by using composition instead of inheritance to limit the access to only a subset of public functions:
```c
  slave_gcm->get_component_with_subset_interface ();
```
If it is not easy to do it in place, open an issue to consider it for later.

A general note here, we need to discuss and document this type of model in an updated code guideline.

File: src/transaction/transaction_master_group_complete_manager.cpp (hunk 6 above)
Comment (ribram) [reply src/transaction/transaction_master_group_complete_manager.cpp]: outdated, it was a merge issue.

File: src/replication/log_consumer.hpp (hunk 1 above)
Comment (alexstanro) [reply src/replication/log_consumer.hpp]: One related discussion was Multiple Inheritance vs. Composition :
http://jira.cubrid.org/browse/CBRD-23218

File: src/transaction/transaction_group_complete_manager.hpp (hunk 5 above)
Comment (alexstanro) [reply src/transaction/transaction_group_complete_manager.hpp]: In single/master, db admin can set such parameter, to simulate actual group commit (see need_wakeup_LFT). Combined with a timer, this min transaction may be used in single/master. I expect the others opinion.

File: src/transaction/transaction_slave_group_complete_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+  //
+  // set_close_info_for_current_group - set close info for current group.
+  //
+  void slave_group_complete_manager::set_close_info_for_current_group (cubstream::stream_position stream_position,
+      int count_expected_transactions)
+  {
+    bool has_group_enough_transactions;
+    /* Can't set close info twice. */
+#if 0
```
[the remark this answers is not in the document]
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.cpp]: Remove and TODO -added.

File: src/transaction/transaction_slave_group_complete_manager.hpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      int get_manager_type () const override;
+
+    protected:
+      /* group_complete_manager methods */
+      bool can_close_current_group () override;
+      void on_register_transaction () override;
+
+    private:
+      static slave_group_complete_manager *gl_slave_group;
+      static cubthread::daemon *gl_slave_group_complete_daemon;
```
[the remark this answers is not in the document]
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.hpp]: No, it said that there are several threads that acts on GCM. Nobody stop the user to use a dedicated thread, more dedicated threads or no dedicated thread, as in your propose.  Complete manager is not singleton.

File: src/connection/server_support.c [hunk 10]
```diff
@@ -2354,7 +2354,7 @@ css_change_ha_server_state (THREAD_ENTRY * thread_p, HA_SERVER_STATE state, bool
 	      ha_Server_state = state;    
 	    }
 
-	  // TODO: investigate the need for log_append
+	  /* append a dummy log record for LFT to wake LWTs up */
```
Comment (ribram) [inline src/connection/server_support.c]: comment is duplicated.

File: src/connection/server_support.c [hunk 11]
```diff
@@ -2741,7 +2741,7 @@ css_process_new_slave (SOCKET master_fd)
 
   auto wp = cubthread::internal_tasks_worker_pool::get_instance ();
   cubthread::get_manager ()->push_task (wp, new_slave_task);
-  // *INDENT-ON*
+  // todo: wait for ha_Server_state to become HA_SERVER_STATE_ACTIVE
```
Comment (ribram) [inline src/connection/server_support.c]: when should this todo be solved? a specific jira issue might be needed if it hasn't one already.

File: src/transaction/transaction_group_complete_manager.hpp (hunk 5 above)
Comment (eseokoh) [reply src/transaction/transaction_group_complete_manager.hpp]: I don't think we can/will have a minimum number of transactions in a group. We should also care for latency of each transaction as well as throughput.

File: src/transaction/transaction_slave_group_complete_manager.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.hpp]: http://jira.cubrid.org/browse/CBRD-23225

File: src/transaction/transaction_group_complete_manager.hpp [hunk 12]
```diff
@@ -47,6 +47,17 @@ namespace cubtx
 			| GROUP_COMPLETE_STARTED | GROUP_COMPLETED)
   };
 
+  //
+  // group completion is the common interface used to control group creation.
+  //
+  class group_completion
```
Comment (ribram) [inline src/transaction/transaction_group_complete_manager.hpp]: It doesn't look a good fit of name and file. Generic group complete manager is not directly linked to stream/stream position.

It should belong to cubreplication namespace and it should be moved to replication/slave/apply related files. One quick action could be moving to slave GCM file, rename it as slave_group_complete_controller. Its name and place can be later reconsidered its with http://jira.cubrid.org/browse/CBRD-23218.

File: src/transaction/transaction_master_group_complete_manager.cpp [hunk 13]
```diff
@@ -128,7 +128,7 @@ namespace cubtx
   {
     if (!is_latest_closed_group_completed ())
       {
-	/* Can't advance to the next group since the current group was not completed yet. */
+	/* Can't advance to the next group since the current group was not committed yet. */
```
Comment (ribram) [inline src/transaction/transaction_master_group_complete_manager.cpp]: old comment looks more accurate.

File: src/transaction/transaction_slave_group_complete_manager.hpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+#ifndef _TRANSACTION_SLAVE_GROUP_COMPLETE_MANAGER_HPP_
+#define _TRANSACTION_SLAVE_GROUP_COMPLETE_MANAGER_HPP_
+
+#include "transaction_group_complete_manager.hpp"
+
+#include "cubstream.hpp"
+#include "log_consumer.hpp"
+#include "thread_daemon.hpp"
+#include "thread_entry_task.hpp"
```
Comment (ribram) [inline src/transaction/transaction_slave_group_complete_manager.hpp]: `#include "thread_entry_task.hpp"` no longer needed here. probably it must be included in cpp file.

File: src/transaction/transaction_slave_group_complete_manager.hpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+      int get_manager_type () const override;
+
+    protected:
+      /* group_complete_manager methods */
+      bool can_close_current_group () override;
+      void on_register_transaction () override;
+
+    private:
+      static slave_group_complete_manager *gl_slave_gcm;
```
Comment (ribram) [inline src/transaction/transaction_slave_group_complete_manager.hpp]: The static instance inside the class misleads reader that class is singleton. To clarify this aspect, gl_slave_gcm/gl_slave_group_complete_daemon and the functions to init/get/finalize them could be moved outside slave_group_complete_manager class (just regular global variables inside cpp files and regular functions init_slave_gcm/get_slave_gcm/final_slave_gcm to be used from other units).

**Comment by radudoros** [comment]
I also noticed that many member funcs from the base class group_complete_manager are possible to be made const (all is_*, get_current_group)

File: src/transaction/transaction_slave_group_complete_manager.hpp (hunk 15 above)
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.hpp]: Ok, I changed initialization, similarly with g_master_node initialization, even if I'm not sure that is the best option.

File: src/transaction/transaction_slave_group_complete_manager.cpp [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+      }
+
+    /* Finally, notify complete. TODO - consider notify log and complete together. Consider complete MVCC case 2. */
+    notify_group_complete ();
+  }
+
+  //
+  // complete_upto_stream_position - complete up to stream position
+  //
+  void slave_group_complete_manager::complete_upto_stream_position (cubstream::stream_position stream_position)
```
[the remark this answers is not in the document]
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.cpp]: Yes.

File: src/transaction/transaction_slave_group_complete_manager.cpp [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+  {
+    if (stream_position <= m_latest_group_stream_position)
```
[the remark this answers is not in the document]
Comment (alexstanro) [reply src/transaction/transaction_slave_group_complete_manager.cpp]: Is same. I can change it as you request since it simplifies the code.

File: src/transaction/transaction_slave_group_complete_manager.hpp [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    private:
+      /* Latest recorded stream position and corresponding id. */
+      id_type m_latest_group_id;
+      std::atomic<cubstream::stream_position> m_latest_group_stream_position;
+
+      /* m_has_latest_group_close_info - true, if stream position and count expected transactions were set. */
+      std::atomic<bool> m_has_latest_group_close_info;
+  };
+
+  void initialize ();
```
Comment (ribram) [inline src/transaction/transaction_slave_group_complete_manager.hpp]: this is part of cubtx namespace and could be in conflict with other generically named functions of same module.

better to add "slave_gcm" in name, e.g.

```c
  void initialize_slave_gcm ();
  void finalize_slave_gcm ();
  slave_group_complete_manager *get_slave_gcm_instance ();

File: src/transaction/transaction_group_complete_manager.hpp (hunk 5 above)
Comment (alexstanro) [reply src/transaction/transaction_group_complete_manager.hpp]: Now, that we know that it will be only a feature of slave GCM, we can move it in slave files.

However, moving is a risky job at this moment, since group mutex (stored in base class) is used to atomic check group state. Also, we don't want to break the actual design.

If we break the consistency, a hang may occur that will be quite difficult to catch.

So, my conclusion is that our focus should be consistency not perfect design, in this moment.

But, since some guys continuously insists, beyond my understanding, I tried to find a proper solution...

File: src/transaction/transaction_group_complete_manager.hpp (hunk 12 above)
Comment (ribram) [reply src/transaction/transaction_group_complete_manager.hpp]: This is my only remaining unresolved comment.

Comment (alexstanro) [reply src/transaction/transaction_group_complete_manager.hpp]: Please clarify with Razvan your suggestion. Moving into slave GCM is not suitable since group_completion is accessed by log_consumer. Renaming again, seems to be an issue without end...

Comment (alexstanro) [reply src/transaction/transaction_group_complete_manager.hpp]: Since for some of us looks important, finally, was moved n a new file without changing name.
